File: solutions/symmetric-tree/symmetric-tree.cpp
// @Title: Symmetric Tree
// @Author: colinjxc
// @Date: 2018-10-19T22:08:17+08:00
// @URL: https://leetcode-cn.com/problems/symmetric-tree


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {    
public:
    bool isSymmetric(TreeNode* root) {
        if (root == NULL) {
            return true;
        }
        return  issym(root->left, root->right);
    }
    
private:
    bool issym(TreeNode* l, TreeNode* r) {
        if (l == NULL && r == NULL) {
            return true;
        }
        
        if (l == NULL || r == NULL) {
            return false;
        }
        
        if (l->val == r->val) {
            return issym(l->left, r->right) && issym(l->right, r->left);
        } else {
            return false;
        }

    }
};

File: solutions/symmetric-tree/symmetric-tree.go
// @Title: Symmetric Tree
// @Author: colinjxc
// @Date: 2021-07-23T18:39:42+08:00
// @URL: https://leetcode-cn.com/problems/symmetric-tree


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }
    return isSym(root.Left, root.Right)
}

func isSym(left, right *TreeNode) bool {
    if left == nil && right ==  nil {
        return true
    } 

    if left == nil || right == nil  {
        return false
    }

    if left.Val != right.Val {
        return false
    } else {
        return isSym(left.Left, right.Right) && isSym(left.Right, right.Left)
    }
}
